src/NHibernate/Transform/QueryAliasToObjectPropertySetter.cs

```diff
@@ -0,0 +1,196 @@
+			{
+				checkMember(property);
+				property.Member.SetValue(resultObj, value);
+				return;
+			}
+			if (_fieldsByNameCaseInsensitive.TryGetValue(alias, out field))
+			{
+				checkMember(field);
+				field.Member.SetValue(resultObj, value);
+				return;
+			}
+
+			throw new PropertyNotFoundException(resultObj.GetType(), alias, "setter");
+
+			void checkMember<T>(NamedMember<T> member) where T : MemberInfo
+			{
+				if (member.Member != null)
+					return;
+
+				if (member.AmbiguousMembers == null || member.AmbiguousMembers.Length < 2)
+					throw new InvalidOperationException($"{nameof(NamedMember<T>.Member)} missing and {nameof(NamedMember<T>.AmbiguousMembers)} invalid.");
+
+				throw new AmbiguousMatchException(
+					$"Unable to find adequate property or field to set on '{member.AmbiguousMembers[0].DeclaringType.Name}' for alias '{alias}', " +
+						$"many {(member.AmbiguousMembers[0] is PropertyInfo ? "properties" : "fields")} matches: {string.Join(", ", member.AmbiguousMembers.Select(m => m.Name))}");
+			}
+		}
+
+		/// <summary>
+		/// Build a <c>QueryAliasToObjectPropertySetter</c> for a supplied type.
+		/// </summary>
+		/// <param name="objType">The type.</param>
+		/// <returns>A <c>QueryAliasToObjectPropertySetter</c>.</returns>
+		public static QueryAliasToObjectPropertySetter MakeFor(System.Type objType)
+		{
+			var bindingFlags = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly;
+			var fields = new List<RankedMember<FieldInfo>>();
+			var properties = new List<RankedMember<PropertyInfo>>();
+			var currentType = objType;
+			var rank = 1;
+			// For grasping private members, we need to manually walk the hierarchy.
+			while (currentType != null && currentType != typeof(object))
+			{
+				fields.AddRange(
+					currentType
+						.GetFields(bindingFlags)
+						.Select(f => new RankedMember<FieldInfo> { Member = f, VisibilityRank = getFieldVisibilityRank(f), HierarchyRank = rank }));
+				properties.AddRange(
+					currentType
+						.GetProperties(bindingFlags)
+						.Where(p => p.CanWrite && p.GetIndexParameters().Length == 0)
+						.Select(p => new RankedMember<PropertyInfo> { Member = p, VisibilityRank = getPropertyVisibilityRank(p), HierarchyRank = rank }));
+				currentType = currentType.BaseType;
+				rank++;
+			}
+
+			var fieldsByNameCaseSensitive = getMapByName(fields, StringComparer.Ordinal);
+			var fieldsByNameCaseInsensitive = getMapByName(fields, StringComparer.OrdinalIgnoreCase);
+			var propertiesByNameCaseSensitive = getMapByName(properties, StringComparer.Ordinal);
+			var propertiesByNameCaseInsensitive = getMapByName(properties, StringComparer.OrdinalIgnoreCase);
+
+			return new QueryAliasToObjectPropertySetter(fieldsByNameCaseSensitive, fieldsByNameCaseInsensitive, propertiesByNameCaseSensitive, propertiesByNameCaseInsensitive);
+
+			int getFieldVisibilityRank(FieldInfo field)
```

Comment:
PascalCase? Microsoft seems to mandate camelCase instead, for exhibiting their local nature.